flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java

```diff
@@ -0,0 +1,340 @@
+
+                for (int binEdgeId = 0; binEdgeId < numBins; binEdgeId++) {
+                    tempBinEdges[binEdgeId] = features[(int) (binEdgeId * width)];
+                }
+                tempBinEdges[numBins] = features[numData - 1];
+
+                // Removes bins who are empty, i.e., the left edge equals to the right edge.
+                Set<Double> edges = new HashSet<>(numBins);
+                for (double edge : tempBinEdges) {
+                    edges.add(edge);
+                }
+
+                binEdges[columnId] = edges.stream().mapToDouble(Double::doubleValue).toArray();
+                Arrays.sort(binEdges[columnId]);
+            }
+        }
+        return binEdges;
+    }
+
+    private static double[][] findBinEdgesWithKMeansStrategy(List<DenseVector> input, int numBins) {
+        int numColumns = input.get(0).size();
+        int numData = input.size();
+        double[][] binEdges = new double[numColumns][numBins + 1];
+        double[] features = new double[numData];
+
+        double[] kMeansCentroids = new double[numBins];
+        double[] sumByCluster = new double[numBins];
+
+        for (int columnId = 0; columnId < numColumns; columnId++) {
+            for (int i = 0; i < numData; i++) {
+                features[i] = input.get(i).get(columnId);
+            }
+            Arrays.sort(features);
+
+            if (features[0] == features[numData - 1]) {
+                LOG.warn("Feature " + columnId + " is constant and the output will all be zero.");
+                binEdges[columnId] = new double[] {Double.MIN_VALUE, Double.MAX_VALUE};
+            } else {
+                // Checks whether there are more than {numBins} distinct feature values in each
+                // column.
+                // If the number of distinct values is less than {numBins + 1}, then we do not need
+                // to conduct KMeans. Instead, we switch to use {@link
+                // KBinsDiscretizerParams#UNIFORM} for bining.
+                Set<Double> distinctFeatureValues = new HashSet<>(numBins + 1);
+                for (double feature : features) {
+                    distinctFeatureValues.add(feature);
+                    if (distinctFeatureValues.size() >= numBins + 1) {
+                        break;
+                    }
+                }
+                if (distinctFeatureValues.size() <= numBins) {
+                    double min = features[0];
+                    double max = features[features.length - 1];
+                    double width = (max - min) / numBins;
+                    binEdges[columnId] = new double[numBins + 1];
+                    binEdges[columnId][0] = min;
+                    for (int edgeId = 1; edgeId < numBins + 1; edgeId++) {
+                        binEdges[columnId][edgeId] = binEdges[columnId][edgeId - 1] + width;
+                    }
+                    continue;
+                } else {
+                    // Conducts KMeans here.
```

Comment:
Would it be better to reuse the logic in the kmeans or online kmeans stages?

Reply:
It may no be that efficient to reuse the logic in `KMeans` class:
- We do kmeans for each input column here, i.e., the input feature is simply a double, not a vector.
- Reusing `Kmeans` class means we introduce extra flink operators, and make the job more complex. Here a toy code for one-dimension kmeans is enough.